sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,752 @@
+  printf("Sending reported property\n");
+
+  // Get the topic used to send a reported property update
+  if (az_failed(
+          rc = az_iot_hub_client_twin_patch_get_publish_topic(
+              &client,
+              reported_property_topic_request_id,
+              reported_property_topic,
+              sizeof(reported_property_topic),
+              NULL)))
+  {
+    printf("Unable to get twin document publish topic, return code %d\n", rc);
+    return rc;
+  }
+
+  // Twin reported properties must be in JSON format. The payload is constructed here.
+  az_json_builder json_builder;
+  if (az_failed(rc = build_reported_property(&json_builder, desired_temp)))
+  {
+    return rc;
+  }
+  az_span json_payload = az_json_builder_get_json(&json_builder);
+
+  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+
+  // Publish the reported property payload to IoT Hub
+  if ((rc = MQTTClient_publish(
+           mqtt_client,
+           reported_property_topic,
+           az_span_size(json_payload),
+           az_span_ptr(json_payload),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish reported property, return code %d\n", rc);
+    return rc;
+  }
+  return rc;
+}
+
+// Parse the desired temperature property from the incoming JSON
+static az_result parse_desired_temperature_property(az_span twin_span, double* parsed_value)
+{
+  az_json_parser jp;
+  az_json_token_member tm;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, twin_span));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(desired_temp_property_name, tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm.token, parsed_value));
+    }
+
+    // else ignore token.
+  }
+
+  return AZ_OK;
+}
+
+// Callback for incoming MQTT messages
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  (void)context;
+
+  if (topicLen == 0)
+  {
+    // The length of the topic if there are one or more NULL characters embedded in topicName,
+    // otherwise topicLen is 0.
+    topicLen = (int)strlen(topicName);
+  }
+
+  printf("Topic: %s\n", topicName);
+
+  az_span topic_span = az_span_init((uint8_t*)topicName, topicLen);
+
+  // Parse the incoming message topic and check which feature it is for
+  az_iot_hub_client_twin_response twin_response;
+  az_iot_hub_client_method_request method_request;
+
+  if (az_succeeded(
+          az_iot_hub_client_twin_parse_received_topic(&client, topic_span, &twin_response)))
+  {
+    printf("Twin Message Arrived\n");
+
+    // Determine what type of incoming twin message this is. Print relevant data for the message.
+    switch (twin_response.response_type)
```

Comment:
>twin_response.response_type [](start = 12, length = 27)

I remember that in case of non successful response, this classification of message type is wrong atleast between Reported Response and get twin response.

Reply:
Sort of. We can't correlate errors with original messages since that would require keeping an internal list of messages, types, and correlation ID's. For now it's the best we can do until/unless we come up with something else.

Reply:
One thing atleast at convenience layer we did is to divide the request id into odd and even, and then use that extra information to decide message response type.

---
In reply to: [448549680](https://github.com/Azure/azure-sdk-for-c/pull/862#discussion_r448549680) [](ancestors = 448549680)

Reply:
That would work if you have control of the mqtt stack and are able to create the messages for the user. In the raw embedded sdk case, that decision is left to the user so we can't

Reply:
I mean request Id for IoThub, not mqtt. When you create any request, you can add your requestId on the topic and send request to IoTHub, which as part of response will be part of response topic. 

I see that you are adding request Id on line 454

---
In reply to: [448569370](https://github.com/Azure/azure-sdk-for-c/pull/862#discussion_r448569370) [](ancestors = 448569370)

Reply:
Ya I think we're talking about the same thing. For the sample we might be able to do that (user code) but we can't do that with sdk code. Maybe an improvement to be added to the samples.